src/datatype.c

```diff
@@ -855,19 +855,27 @@ JL_DLLEXPORT jl_value_t *jl_new_structt(jl_datatype_t *type, jl_value_t *tup)
     if (type->instance != NULL) {
         for (size_t i = 0; i < na; i++) {
             jl_value_t *ft = jl_field_type(type, i);
-            jl_value_t *fi = jl_get_nth_field(tup, i);
-            if (!jl_isa(fi, ft))
-                jl_type_error("new", ft, fi);
+            jl_datatype_t *et = jl_field_type(jl_typeof(tup), i);
+            assert(!jl_is_uniontype(et));
```

Comment:
it must be a concrete datatype and cannot have `isptr` fields. Plus, loading all fields must also be allocation free (singleton). The `jl_isa` test should be trivial and can be written as `et == fi`.

Reply:
What is `it`? `typeof(tup)`, `type`?

Reply:
Ah, you mean `type`, because we checked that `type->instance` is non-NULL. Fair point.

Reply:
On the other hand, do we want to hardcode that. E.g. doe we ever want to give singleton status to things like:
```
struct Foo
           x::Type{Int}
       end
```
?

Reply:
Meh, I'll add enough asserts that hopefully one of them will fire if it ever comes to this.

Reply:
By the time that could eventually happen, the code here would seem to be potentially wrong anyways, so 🤷‍♂️. Might as well do the simple, fast thing right now.